### PR TITLE
Textanswer visibility info

### PR DESCRIPTION
**Pull request by janno42**
Explain to students on the voting page, to which people their text answers will be shown after publishing.

### CODE REVIEW

File: evap/student/tests/test_views.py [hunk 1]
```diff
@@ -178,11 +178,11 @@ def test_user_cannot_vote_multiple_times(self):
     def test_user_cannot_vote_for_themselves(self):
         response = self.app.get(self.url, user=self.contributor1, status=200)
 
-        for contributor, _, _, _ in response.context['contributor_form_groups']:
+        for contributor, _, _, _, _ in response.context['contributor_form_groups']:
```
Comment (He3lixxx) [inline evap/student/tests/test_views.py]: double underscore maybe?

File: evap/student/tests/test_views.py [hunk 2]
```diff
@@ -178,11 +178,11 @@ def test_user_cannot_vote_multiple_times(self):
...
             self.assertNotEqual(contributor, self.contributor1, "Contributor should not see the questionnaire about themselves")
 
         response = self.app.get(self.url, user=self.voting_user1, status=200)
-        self.assertTrue(any(contributor == self.contributor1 for contributor, _, _, _ in response.context['contributor_form_groups']),
+        self.assertTrue(any(contributor == self.contributor1 for contributor, _, _, _, _ in response.context['contributor_form_groups']),
```
Comment (He3lixxx) [inline evap/student/tests/test_views.py]: double underscore maybe?

File: evap/results/views.py [hunk 3]
```diff
@@ -230,15 +230,23 @@ def user_can_see_text_answer(user, represented_users, text_answer, view):
...
         return contributor == user
 
+    # NOTE: when changing this behavior, make sure all changes are also reflected in results.tools.textanswers_visible_to
+    # and in results.tests.test_tools.TestTextAnswerVisibilityInfo
     if text_answer.is_published:
+        # text answers about responsible contributors can only be seen by the users themselves and by their delegates
+        # they can not be seen by other responsible contributors
         if text_answer.contribution.responsible:
-            return contributor == user or user in contributor.delegates.all()
+            return contributor in represented_users
```
Comment (karyon) [inline evap/results/views.py]: please check again, i'm pretty sure this is not equivalent. previously it was `user in contributor.represented_users`, now it is `contributor in user.represented_users`. if the code in the PR is currently incorrect, please make a test demonstrating the problem.

File: evap/student/templates/student_vote_questionnaire_group.html [hunk 4]
```diff
@@ -43,11 +43,22 @@
...
                             </td>
                         {% else %}
-                            <td class="vote-text-answer">
+                            <td class="vote-text-answer d-flex">
                                 <div class="vote-inputs{% if preview %} preview{% endif %}">
                                     {{ field.errors }}
                                     <textarea id="{{ field.id_for_label }}" name="{{ field.name }}">{{ field.value|default_if_none:"" }}</textarea>
                                 </div>
+                                <div>
+                                    <span class="far fa-eye ml-1 icon-gray" data-toggle="tooltip" data-placement="left" title="{% trans 'After publishing, this text answer can be seen by:' %}
```
Comment (karyon) [inline evap/student/templates/student_vote_questionnaire_group.html]: begin string on newline and then indent plz

File: evap/student/templates/student_vote_questionnaire_group.html [hunk 5]
```diff
@@ -43,11 +43,22 @@
...
+                                    {% for contributor in textanswers_visible_to.0 %}
+                                        {{ contributor.full_name }}{% if not forloop.last or textanswers_visible_to.1|length > 0 %}, {% endif %}
+                                    {% endfor %}
+                                    {% if textanswers_visible_to.1|length > 0 %}
+                                        {% blocktrans count delegate_count=textanswers_visible_to.1|length %}{{ delegate_count }} delegate{% plural %}{{ delegate_count }} delegates{% endblocktrans %}
```
Comment (karyon) [inline evap/student/templates/student_vote_questionnaire_group.html]: can you make textanswers_visible_to a namedtuple so we don't have to use `.1`?

File: evap/results/views.py (hunk 3 above)
Comment (karyon) [reply evap/results/views.py]: you're right, it's correct.

File: evap/results/tests/test_tools.py [hunk 6]
```diff
@@ -190,3 +192,106 @@ def test_result_calculation_with_no_contributor_rating_question_does_not_fail(se
...
+            if user_can_see_text_answer(user, represented_users, self.contributor_course_textanswer, 'full'):
+                if user_can_see_text_answer(user, [user], self.contributor_course_textanswer, 'full'):
+                    users_seeing_contributor_course_contribution[0].add(user)
+                else:
+                    users_seeing_contributor_course_contribution[1].add(user)
+            if user_can_see_text_answer(user, represented_users, self.contributor_all_textanswer, 'full'):
+                if user_can_see_text_answer(user, [user], self.contributor_all_textanswer, 'full'):
+                    users_seeing_contributor_all_contribution[0].add(user)
+                else:
+                    users_seeing_contributor_all_contribution[1].add(user)
```
Comment (karyon) [inline evap/results/tests/test_tools.py]: can we have something... more generic? i mean we have six contributions, each of those has a comment, the result of `textanswers_visible_to`, and then a bunch of users who can or can not see it. i have the feeling all this should be possible with a bunch of loops.

File: evap/results/tests/test_tools.py [hunk 7]
```diff
@@ -190,3 +192,106 @@ def test_result_calculation_with_no_contributor_rating_question_does_not_fail(se
...
+        self.assertEqual(len(responsible1_contribution_delegates), len(users_seeing_responsible1_contribution[1]))
+        self.assertEqual(len(responsible1_contribution_delegates), 2)  # delegate1, contributor_course
+        self.assertEqual(len(responsible2_contribution_delegates), len(users_seeing_responsible2_contribution[1]))
+        self.assertEqual(len(responsible2_contribution_delegates), 0)
+        self.assertEqual(len(contributor_own_contribution_delegates), len(users_seeing_contributor_own_contribution[1]))
+        self.assertEqual(len(contributor_own_contribution_delegates), 2)  # delegate1, contributor_course
+        self.assertEqual(len(contributor_course_contribution_delegates), len(users_seeing_contributor_course_contribution[1]))
+        self.assertEqual(len(contributor_course_contribution_delegates), 2)  # delegate1, delegate2
+        self.assertEqual(len(contributor_all_contribution_delegates), len(users_seeing_contributor_all_contribution[1]))
+        self.assertEqual(len(contributor_all_contribution_delegates), 2)  # delegate1, contributor_course
```
Comment (karyon) [inline evap/results/tests/test_tools.py]: try to compare sets instead of counts. or maybe even lists because you specified the ordering in the code.

Comment (janno42) [reply evap/results/tests/test_tools.py]: these will stay counts when not returning the delegates but only a count in results/tools.py, right?

Comment (karyon) [reply evap/results/tests/test_tools.py]: yes. another thought, maybe do `self.assertTrue(len(general_contribution_delegates) == len(users_seeing_general_contribution[1]) == 2)`?

File: evap/results/tests/test_tools.py [hunk 8]
```diff
@@ -190,3 +192,74 @@ def test_result_calculation_with_no_contributor_rating_question_does_not_fail(se
...
+    def test_textanswer_visibility_info(self):
+        contributions = [
+            self.general_contribution, self.responsible1_contribution, self.responsible2_contribution,
+            self.contributor_own_contribution, self.contributor_course_contribution, self.contributor_all_contribution
+        ]
+        textanswers = [
+            self.general_contribution_textanswer, self.responsible1_textanswer, self.responsible2_textanswer,
+            self.contributor_own_textanswer, self.contributor_course_textanswer, self.contributor_all_textanswer
+        ]
+        visible_to = [textanswers_visible_to(contribution) for contribution in contributions]
```
Comment (karyon) [inline evap/results/tests/test_tools.py]: you can remove the `contributions` list, use `len(textanswers)` everywhere, and access textanswer.contribution here.

**Pull request merged**